File: src/offer5.Rplace_spaces/replace_spaces.cpp
class Solution {
public:
	void replaceSpace(char *str,int length) {
        if(str == nullptr || length <=0) return;
        int count = 0;
        for(int i = 0; i < length; i++){
            if (str[i] == ' '){
                count++;
            }
        }
        int newlen = length + 2*count;
        int ll = length - 1;
        int rr = newlen - 1;
        while(ll < rr && ll >= 0 && rr >= 0){
            if(str[ll] == ' '){
                str[rr--] = '0';
                str[rr--] = '2';
                str[rr--] = '%';
            }
            else{
                str[rr--] = str[ll];
            }
            ll--;
        }
	}
};

File: src/offer05.Rplace_spaces/replace_spaces.java
public class Solution {
    public String replaceSpace(StringBuffer str) {
        int spaceNum = 0;
        for(int i = 0; i < str.length(); i++){
            if(str.charAt(i) == ' '){
                spaceNum++;
            }
        }
        int oldIndex = str.length() - 1;
        int newlen = str.length() + 2 * spaceNum;
        int newIndex = newlen - 1;
        str.setLength(newlen);
        while(oldIndex >= 0 && newIndex >= 0){
            if(str.charAt(oldIndex) == ' '){
                str.setCharAt(newIndex--, '0');
                str.setCharAt(newIndex--, '2');
                str.setCharAt(newIndex--, '%');
            }else{
                str.setCharAt(newIndex--, str.charAt(oldIndex));
            }
            oldIndex--;
        }
        return str.toString();
    }
}